core/arch/instr.h

```diff
@@ -578,12 +617,30 @@ INSTR_INLINE
 void *
 instr_get_note(instr_t *instr);
 
+DR_API
+INSTR_INLINE
+/**
+ * Finds the value of a user-controlled note field in \p instr.
+ * \note Important: is also used when emitting for targets that are other
+ * instructions.  Thus it will be overwritten when calling instrlist_encode()
+ * or instrlist_encode_to_copy() with \p has_instr_jmp_targets set to true.
+ * \note The note field is copied (shallowly) by instr_clone().
+ */
+note_t *
+instr_find_note(instr_t *instr, enum note_type_id type);
```

Comment:
Rather than a list of notes in the app instruction, our approach has been to instead have each note generator add its own label instruction that contains its identifier and any additional data.  So the emulation sequence would be bounded by a start label and an end label, and the start label would contain addditional information in its data fields.

The note field would just contain an integer identifier, mediated by drmgr_reserve_note_range().  The data would be in fields provided by instr_get_label_data_area().

We would also prefer to have this in drmgr rather than the core: have drmgr specify a note identifier value that indicates a sequence of instructions that should be skipped by observational clients and whose data contains a specified format (your emulation_state_t) including a pointer to original instructions which should be examined instead.

Reply:
This solution looks good but I have a question. The only way to check if an instr is a label is to use instr_is_label() which only works if the label is a meta instruction. This means that the instrumentation client loop has to distinguish between app instrs and meta instrs and becomes a bit more complicated. For example, the inscount.cpp loop would look like:

    for (instr = instrlist_first_app(bb), num_instrs = 0;                                             
         instr != NULL;                                                                               
         instr = instr_get_next(instr)) {                                                             
        if (instr_is_label(instr)) {                                                                  
            if (((ptr_uint_t)instr_get_note(instr)) == DRMGR_NOTE_START_EMULATION)                    
                fprintf(stderr, "start->\n");                                                         
            if (((ptr_uint_t)instr_get_note(instr)) == DRMGR_NOTE_STOP_EMULATION)                     
                fprintf(stderr, "<-stop\n");                                                          
        }                                                                                             
        if (!instr_is_app(instr))                                                                     
            continue;
      . . .
Is there a way of using the labelling method without changing the existing client loops, i.e. to deal with a meta label instr and the app instrs?

Reply:
The loop would normally use `instr_get_next_app`.  Since label instructions are nearly always added as meta, they would be skipped by the iterator and no explicit check is needed.